client/sysinfo/src/sysinfo.rs

```diff
@@ -32,14 +37,144 @@ use std::{
+
+const KIBIBYTE: f64 = 1024.0;
+
+impl Throughput {
+	/// `f64` kibibyte/s to byte/s.
+	pub fn from_kibs(kibs: f64) -> Throughput {
+		Throughput(kibs * KIBIBYTE)
+	}
+
+	/// `f64` mebibyte/s to byte/s.
+	pub fn from_mibs(mibs: f64) -> Throughput {
+		Throughput(mibs * KIBIBYTE * KIBIBYTE)
+	}
+
+	/// `f64` gibibyte/s to byte/s.
+	pub fn from_gibs(gibs: f64) -> Throughput {
+		Throughput(gibs * KIBIBYTE * KIBIBYTE * KIBIBYTE)
+	}
+
+	pub fn as_f64(&self) -> f64 {
+		self.0
+	}
+
+	/// [`Self`] as number of kibibyte/s.
+	pub fn as_kibs(&self) -> f64 {
+		self.0 / KIBIBYTE
+	}
+
+	/// [`Self`] as number of mebibyte/s.
+	pub fn as_mibs(&self) -> f64 {
+		self.0 / (KIBIBYTE * KIBIBYTE)
+	}
+
+	/// [`Self`] as number of gibibyte/s.
+	pub fn as_gibs(&self) -> f64 {
+		self.0 / (KIBIBYTE * KIBIBYTE * KIBIBYTE)
+	}
+
+	/// Normalizes [`Self`] to use the larges unit possible.
+	fn normalize(&self) -> (f64, &'static str) {
+		let bs = self.0;
+
+		if bs >= KIBIBYTE * KIBIBYTE * KIBIBYTE {
+			(self.as_gibs(), "GiBs")
+		} else if bs >= KIBIBYTE * KIBIBYTE {
+			(self.as_mibs(), "MiBs")
+		} else {
+			(self.as_kibs(), "KiBs")
+		}
+	}
+}
+
+impl fmt::Display for Throughput {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		let (value, unit) = self.normalize();
+		write!(f, "{:.2?} {}", value, unit)
+	}
+}
+
+pub fn serialize_throughput<S>(t: &Throughput, serializer: S) -> Result<S::Ok, S::Error>
+where
+	S: Serializer,
+{
+	// NOTE I will replace 4 with the actual length.
+	let mut map = serializer.serialize_map(Some(4))?;
+	let (value, unit) = t.normalize();
+	map.serialize_entry(unit, &value)?;
+	map.end()
+}
+
+pub fn serialize_throughput_as_mibs<S>(t: &Throughput, serializer: S) -> Result<S::Ok, S::Error>
+where
+	S: Serializer,
+{
+	let mut map = serializer.serialize_map(Some(4))?;
+	map.serialize_entry("MiBs", &t.as_mibs())?;
+	map.end()
+}
+
+pub fn serialize_throughput_option_as_mibs<S>(
+	maybe_t: &Option<Throughput>,
+	serializer: S,
+) -> Result<S::Ok, S::Error>
+where
+	S: Serializer,
+{
+	let mut map = serializer.serialize_map(Some(4))?;
+	if let Some(t) = maybe_t {
+		map.serialize_entry("MiBs", &t.as_mibs())?;
+	}
+	map.end()
+}
+
+struct ThroughputVisitor;
+
+impl<'de> Visitor<'de> for ThroughputVisitor {
+	type Value = Throughput;
+
+	fn expecting(&self, formatter: &mut Formatter) -> fmt::Result {
+		formatter.write_str("A f64")
+	}
+
+	fn visit_map<M>(self, mut access: M) -> Result<Self::Value, M::Error>
+	where
+		M: MapAccess<'de>,
+	{
+		let (key, value): (&str, f64) =
+			if let Some((key, value)) = access.next_entry()? { (key, value) } else { todo!() };
+		if key == "KiBs" {
+			return Ok(Throughput(value * KIBIBYTE))
+		} else if key == "MiBs" {
+			return Ok(Throughput(value * KIBIBYTE * KIBIBYTE))
+		} else {
+			return Ok(Throughput(value * KIBIBYTE * KIBIBYTE * KIBIBYTE))
+		}
+	}
+}
+
+impl<'de> Deserialize<'de> for Throughput {
+	fn deserialize<D>(deserializer: D) -> Result<Throughput, D::Error>
```

Comment:
I believe that it is the best to have the deserialization be implemented for `Throughput` instead of a stand-alone function, since the deserialization logic is always the same.

Reply:
My reasoning here follows something like this:

1. The natural way to serialize/deserialize this is to just do it in full precision. (Which is what a `#[derive]`'d impl would do.)
2. Telemetry doesn't use this and needs a different serialization format. (Not full precision; in MB/s)
3. The code in benchmarking-cli needs a different serialization format. (Wrapped with the unit for readability)
4. Since nothing uses the "natural" serialization format and every piece of code expects a different serialization format in isn't appropriate to define it on the type itself (it's unnecessary, and one might use it by mistake)
5. If the serialization isn't defined on the type itself then neither the deserialization should.

So basically just move the `Deserialize` impl into a standalone function, and move it into the benchmarking-cli where it's actually used.